### DISCUSSION TITLE
Collection's type definition is missing declarations for `$` and `get`

### DISCUSSION THREAD

**Discussion by bbuck** [mikro-orm/mikro-orm#4251] [General]
**Describe the bug**
I'm trying to write tests for code that uses `Collection#init` and I'm encountering an issue mocking the return value (with jest). The implementation of `init` returns the instance of the collection but if I try and mock it to do this I get the below error in TypeScript because the typings for `Collection` are missing definitions for `$` and `get`.

It's possible it's because of the way these properties are defined: https://github.com/mikro-orm/mikro-orm/blob/v5.6.15/packages/core/src/entity/Collection.ts#L433-L436

**Stack trace**
```
 Type 'Collection<Entity, object>' is missing the following properties from type 'LoadedCollection<Loaded<Entity, string>>': $, get
```

**To Reproduce**
1. Have a jest configuration with TypeScript
2. Have a defined entity with a `Collection` field
3. Spy on, and mock the collection, `jest.spyOn(myEntity.collection, 'init').mockReturnValue(Promise.resolve(myEntity.collection)`

**Expected behavior**
No compilation error as all the fields necessary to match the `LoadedCollection` interface are present on `Collection`.

**Versions**

| Dependency | Version |
| - | - |
| node | 18.15 |
| typescript | 4.9.5 |
| mikro-orm | 5.6.15 |
| mikro-orm/postgresql | 5.6.15 |

**Comment by B4nan** [comment]
> It's possible it's because of the way these properties are defined: [v5.6.15/packages/core/src/entity/Collection.ts#L433-L436](https://github.com/mikro-orm/mikro-orm/blob/v5.6.15/packages/core/src/entity/Collection.ts#L433-L436)

Have you seen [type-safe relations](https://mikro-orm.io/docs/type-safe-relations#loaded-type) section of the docs? This is expected, they are defined on prototype only, and added to the type level conditionally - only for loaded properties.

> No compilation error as all the fields necessary to match the LoadedCollection interface are present on Collection.

This would totally break the feature, this difference is there on purpose to force you to initialize such collections before you work with them. Note that tha particular line you mentioned has no effect on TS, it only modifies class prototype - your problem is elsewhere, you want to put a regular `Collection` in place of `LoadedCollection`. 

If you share your test case, I can try to help you with that. Note that I would generally discourage mocking of the ORM internals. But I need to see complete reproduction so its clear what you are doing.

**Reply by bbuck** [reply]
Ultimately the debate has come down to this, we're writing controller tests and there has been some pushback on including actual database access. I would imagine that it's not feasible to test data operations across request (creation of records, updating, deletion, etc...) with mikro-orm without a connection to a database? 

This is why I was pursuing such an aggressive mocking pattern. I don't like mocking dependencies if I can avoid myself.

The test case that I was working on was simply returning the children of an entity via a collection relationship. The controller body was no more than:

```ts
describe("ChildEntityController", () => {
  let orm: MikroORM<PostgreSqlDriver>;
  let em: EntityManager;
  let controller: EntityController;

  beforeAll(async () => {
    orm = await MikroORM.init<PostgreSqlDriver>(
      defineConfig({
        type: "postgresql",
        connect: false,
        entities: [Entity, ChildEntity],
      })
    );
  });

  beforeEach(async () => {
    // new identity map per test
    em = orm.em.fork() as EntityManager;

    const module = await Test.createTestingModule({
      providers: [
        {
          provide: getRepositoryToken(Entity),
          useFactory: jest.fn(() => ({})),
        },
        {
          provide: getRepositoryToken(ChildEntity),
          useFactory: jest.fn(() => ({})),
        },
        {
          provide: SqlEntityManager,
          useFactory: jest.fn(() => ({})),
        },
      ],
      controllers: [ChildEntityController],
    }).compile();

    controller = module.get<ChildEntityController>(ChildEntityController);
  });

  // controller method (nest.js controller)
  // index(
  //   @Param("entityId", LoadEntityPipe())
  //   entity: Entity) {
  //   entity.childEntities.init();

  //   return [...entity.childEntities];
  // }

  describe("index", () => {
    it("returns an empty array if the entity has no child entities", async () => {
      const entity = new Entity();
      em.persist(entity);

      const result = await controller.index(entity);

      expect(result).toEqual([]);
    });
  });
});
```

I believe, in discussion with teammates, that the core issue with this approach is that I'm forking the global EM and it's not the same one that gets handed down to the controller. In this case I should probably be using `MikroORMModule`, but ultimately I have to test creation and deletion endpoints so I'm not sure mocking is even the best approach. Unless I can test all these things without connecting to a database.
